src/core/lib/gpr/alloc.cc

```diff
@@ -33,8 +34,56 @@ static void* zalloc_with_gpr_malloc(size_t sz) {
   return p;
 }
 
-static gpr_allocation_functions g_alloc_functions = {malloc, zalloc_with_calloc,
-                                                     realloc, free};
+#ifndef NDEBUG
+static constexpr bool is_power_of_two(size_t value) {
+  // 2^N =     100000...000
+  // 2^N - 1 = 011111...111
+  // (2^N) && ((2^N)-1)) = 0
+  return (value & (value - 1)) == 0;
+}
+#endif
+
+static void* platform_malloc_aligned(size_t size, size_t alignment) {
+#if defined(GPR_HAS_ALIGNED_ALLOC)
+  GPR_DEBUG_ASSERT(is_power_of_two(alignment));
+  size = GPR_ROUND_UP_TO_ALIGNMENT_SIZE(size, alignment);
+  void* ret = aligned_alloc(alignment, size);
+  GPR_ASSERT(ret != nullptr);
+  return ret;
+#elif defined(GPR_HAS_ALIGNED_MALLOC)
+  GPR_DEBUG_ASSERT(is_power_of_two(alignment));
+  void* ret = _aligned_malloc(size, alignment);
+  GPR_ASSERT(ret != nullptr);
+  return ret;
+#elif defined(GPR_HAS_POSIX_MEMALIGN)
+  GPR_DEBUG_ASSERT(is_power_of_two(alignment));
+  GPR_DEBUG_ASSERT(alignment % sizeof(void*) == 0);
+  void* ret = nullptr;
+  GPR_ASSERT(posix_memalign(&ret, alignment, size) == 0);
+  return ret;
+#else
```

Comment:
Please do not use this pattern. No code should ever be behind a `#else`. Instead, we should have a `#error`. The "default" case should be another `#define` that we can toggle. For instance, due to #18905, we need to be able to toggle the default case for backward compatibility purpose when shipping binary files.

Reply:
@nicolasnoble  What would you recommend instead? The high level goal here is to use a fallback method if no platform specific mode is present. https://github.com/grpc/grpc/pull/18965